modules/odr_caption/odr_caption/outlines/caption.py

```diff
@@ -0,0 +1,78 @@
+import json
+import re
+from PIL import Image
+from odr_caption.outlines.vision import load_images
+from typing import List, Dict, Any
+from pathlib import Path
+import os
+from outlines.generate.api import GenerationParameters, SamplingParameters
+from tqdm import tqdm
+
+
+class CustomEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, str):
+            # Remove escaped unicode characters
+            return re.sub(r"\\u[0-9a-fA-F]{4}", "", obj)
+        return super().default(obj)
+
+
+def process_images_and_save(
+    images: List[Image.Image], instruction: str, output_file: str, generator
+):
+    results = []
+
+    for filename, image in images.items():
+        result = generator(instruction, [image])
+        result_dict = {filename: result.model_dump()}
+        results.append(result_dict)
+
+        os.makedirs(Path(output_file).parent, exist_ok=True)
+        # Save the current results to the JSON file
+        with open(output_file, "w", encoding="utf-8") as f:
+            json.dump(results, f, indent=2, ensure_ascii=False, cls=CustomEncoder)
+
+        print(f"Processed image {filename} and saved results")
+        print(result_dict)
+
+    print(f"All results saved to {output_file}")
+    return results
+
+
+pixtral_instruction = """
+<s>[INST]
+<Task>You are a structured image analysis agent. Generate comprehensive tag list, caption, and dense caption for an image classification system.</Task>
+<TagCategories requirement="You should generate a minimum of 1 tag for each category." confidence="Confidence score for the tag, between 0 (exclusive) and 1 (inclusive).">
+- Entity : The content of the image, including the objects, people, and other elements.
+- Relationship : The relationships between the entities in the image.
+- Style : The style of the image, including the color, lighting, and other stylistic elements.
+- Attribute : The most important attributes of the entities and relationships in the image.
+- Composition : The composition of the image, including the arrangement of elements.
+- Contextual : The contextual elements of the image, including the background, foreground, and other elements.
+- Technical : The technical elements of the image, including the camera angle, lighting, and other technical details.
+- Semantic : The semantic elements of the image, including the meaning of the image, the symbols, and other semantic details.
+<Examples note="These show the expected format as an abstraction.">
+{
+  "tags_list": [
+    {
+      "tag": "subject 1",
+      "category": "Entity",
+      "confidence": 0.98
+    },
+    {
+      "tag": "subject 2",
+      "category": "Entity",
+      "confidence": 0.95
+    },
+    {
+      "tag": "subject 1 runs from subject 2",
+      "category": "Relationship",
+      "confidence": 0.90
+    },
```

Comment:
There is a missing closing ']' bracket in the tags_list example here.

Reply:
nice catch! Luckily we are not relying on the llm to generate the structure so unlikely to cause an issue, but will fix regardless.